Fix/consistent path normalization

AlePouroullis opened the pull request:
Most of the diffs are from rewriting/adding/removing tests. 

The core logic that changed was the handling of paths in the `.call()` and `.log()` operations when using local files.

AlePouroullis committed:
fix: improve path handling in sync client pull operation

AlePouroullis committed:
Improve and make path handling more consistent in call and log overloads

AlePouroullis committed:
refactor(tests): update normalize_path test to cover strip_extension parameter

AlePouroullis committed:
fix: improve path validation for SDK calls

- Enhance safety in path extension handling
- Fix path validation tests to work with actual error messages
- Prioritize extension validation over other path format issues
- Ensure consistent error handling across prompt and agent clients

AlePouroullis committed:
feat: add path utils for centralized path validation

AlePouroullis committed:
Further refined path handling and added more tests

AlePouroullis committed:
refactor: use pytest tmp_path fixture to isolate test file operations

This change:
- Replaces hardcoded directories with pytest's tmp_path fixture
- Eliminates filesystem pollution between tests
- Enables safe test parallelization
- Improves path validation and error messages
- Removes redundant cleanup code
- Makes tests follow pytest best practices

peadaroh commented on src/humanloop/cli/__main__.py:
```diff
@@ -81,6 +81,7 @@ def get_client(
     """
     if not api_key:
         api_key = load_api_key(env_file)
+    print(api_key)
```
Is this intentional?

AlePouroullis replied:
I had that there for debugging (my environment variables didn't seem to update after reloading window).
Good catch.

peadaroh commented on src/humanloop/cli/__main__.py:
```diff
@@ -218,6 +220,7 @@ def pull(
 
     Currently only supports syncing Prompt and Agent files. Other file types will be skipped."""
     client = get_client(api_key, env_file, base_url)
+    # Although pull() is available on the Humanloop client, we instantiate SyncClient separately as we need to control its log level
```
Could you elaborate on why we need to control log level? 

Does this have any implication on other things that happen in constructor related to OTel?

peadaroh commented on src/humanloop/client.py:
```diff
@@ -431,6 +431,11 @@ def pull(self, path: Optional[str] = None, environment: Optional[str] = None) ->
+        Important note about paths:
+        - Paths should not contain leading or trailing slashes
+        - When specifying a file path, include the file extension (`.prompt` or `.agent`)
+        - When specifying a directory path, do not include a trailing slash
```
> When specifying a directory path, do not include a trailing slash

What happens if they do? Thought this would be something we should be able to santize

AlePouroullis replied:
I realize this is incorrect here. `path` in the `pull()` operation is more liberal in terms of format—it normalizes/cleans up silently. `path` in `call()` and `log()` are strict, performing validation but no silent sanitization.

peadaroh commented on src/humanloop/overload.py:
```diff
@@ -87,45 +89,65 @@ def _handle_tracing_context(kwargs: Dict[str, Any], client: Any) -> Dict[str, An
+        try:
+            parts = path.rsplit(".", 1)
+            path_without_extension = parts[0] if len(parts) > 0 else path
+        except Exception:
```
Can you not be more specific on what type of exception you are defending against here?

harry-humanloop replied:
+1 for avoiding bare exceptions - and this should've had a comment.

(think this can be avoided with path.stem as linked above)

peadaroh commented on src/humanloop/overload.py:
```diff
@@ -173,7 +195,11 @@ def _overload_log(self: Any, sync_client: Optional[SyncClient], use_local_files:
-        kwargs = _handle_local_files(kwargs, self, sync_client, use_local_files)
+        if use_local_files and _get_file_type_from_client(self) in SyncClient.SERIALIZABLE_FILE_TYPES:
+            if sync_client is None:
+                logger.error("sync_client is None but client has call method and use_local_files=%s", use_local_files)
```
This error message is hard to parse

peadaroh commented on src/humanloop/path_utils.py:
```diff
@@ -0,0 +1,22 @@
+
+    # Check if path is absolute
+    if path_obj.is_absolute():
+        raise HumanloopRuntimeError("Absolute paths are not supported. Ensure the ")
```
Ensure the what?

AlePouroullis replied:
I was meaning to modify this function to accept a base_dir that could be used to inform the user that their path should be relative to it. However, this only makes sense in the context of the sync client, not a generic util.

On review, I realize that this check is not necessary, and in fact erroneous: shouldn't be doing validation in a normalization function. And redundant, since we strip the leading/trailing slashes prior.

peadaroh commented on tests/custom/integration/test_local_file_operations.py:
```diff
@@ -0,0 +1,314 @@
+            )
+
+
+# def test_agent_local_file_usage(
```
Why is this commented out?

peadaroh commented on tests/custom/integration/test_sync.py:
```diff
@@ -47,165 +32,71 @@ def test_pull_basic(
+        humanloop_client.pull(environment="invalid_environment")
+
+
+# def test_pull_with_environment(
```
why commented out?

AlePouroullis committed:
test: fix type errors in local file operations test by using proper ChatMessageParams type

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -87,45 +89,65 @@ def _handle_tracing_context(kwargs: Dict[str, Any], client: Any) -> Dict[str, An
-    if not use_local_files or "path" not in kwargs or sync_client is None:
-        return kwargs
-
+    """Handle local file loading."""
```
docstring needs elaboration - what does "handle" mean? what's the returned dict? what's "kwargs"?

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -87,45 +89,65 @@ def _handle_tracing_context(kwargs: Dict[str, Any], client: Any) -> Dict[str, An
+    path = kwargs["path"]
+
+    # First check for path format issues (absolute paths or leading/trailing slashes)
+    normalized_path = path.strip("/")
```
I'm surprised this is necessary. why are we doing something so manually?
also: https://docs.python.org/3/library/os.html#os.sep

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -87,45 +89,65 @@ def _handle_tracing_context(kwargs: Dict[str, Any], client: Any) -> Dict[str, An
+    # Then check for file extensions
+    if sync_client.is_file(path):
+        try:
+            parts = path.rsplit(".", 1)
```
https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem

should avoid doing this path string processing imo

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -87,45 +89,65 @@ def _handle_tracing_context(kwargs: Dict[str, Any], client: Any) -> Dict[str, An
+        except Exception:
+            path_without_extension = path
+
+        raise HumanloopRuntimeError(
```
am i misreading the indentation here?
this looks like it always raises an error if `sync_client.is_file(path)`

AlePouroullis committed:
docs(cli): clarify SyncClient log level control and OpenTelemetry isolation

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -151,11 +173,11 @@ def _overload_log(self: Any, sync_client: Optional[SyncClient], use_local_files:
-        if _get_file_type_from_client(self) in ["prompt", "agent"]:
+        if use_local_files and _get_file_type_from_client(self) in SyncClient.SERIALIZABLE_FILE_TYPES:
             if sync_client is None:
                 logger.error("sync_client is None but client has log method and use_local_files=%s", use_local_files)
```
what does "client has log method" mean?

harry-humanloop commented on src/humanloop/path_utils.py:
```diff
@@ -0,0 +1,16 @@
+        normalized = str(path_obj)
+
+    # Normalize separators
+    return "/".join(part for part in normalized.replace("\\", "/").split("/") if part)
```
is this https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.as_posix

harry-humanloop commented on src/humanloop/path_utils.py:
```diff
@@ -0,0 +1,16 @@
+
+
+def normalize_path(path: str, strip_extension: bool = False) -> str:
+    """Normalize a path to the standard API format: path/to/resource, where resource can be a file or a directory."""
```
should be expanded with what this is for, and a few examples imo.

i can't tell why we strip leading/trailing slashes and whether that's correct.

harry-humanloop commented on tests/custom/unit/test_path_utils.py:
```diff
@@ -0,0 +1,28 @@
+def test_normalize_path():
+    """Test path normalization functionality."""
+    # GIVEN various file paths with different formats
+    test_cases = [
```
should this use pytest.mark.parametrize? https://docs.pytest.org/en/stable/how-to/parametrize.html#parametrize-basics

would give the three items in the tuples better variable names

AlePouroullis committed:
refactor: simplified path processing to use pathlib where possible

AlePouroullis committed:
docs: improve comment clarity

AlePouroullis committed:
test: improve path normalization tests with parametrize and edge cases

AlePouroullis committed:
docs: clarify pull_file docstring with failure example

AlePouroullis committed:
docs: expand normalize_path docstring with usage and examples

- Add detailed explanation of function's purpose and usage contexts\n- Document rationale for stripping leading/trailing slashes\n- Add comprehensive examples for different path formats\n- Reference SyncClient.pull usage for context

AlePouroullis committed:
Merge branch 'master' of github.com:humanloop/humanloop-python into fix/consistent-path-normalization

AlePouroullis committed:
refactor: SyncClient -> FileSyncer

AlePouroullis committed:
fix(sync): Convert base_dir to string in FileSyncer fixture

AlePouroullis committed:
fix(dosc): correct logger namespace reference

AlePouroullis committed:
docs: Improve error messages and comments in FileSyncer

AlePouroullis committed:
refactor(test): use pytest.mark.parametrize for path validation tests

Consolidate the three similar test loops (extension paths, slash paths, and combined paths)
into a single parametrized test. This reduces code duplication while making test cases
more maintainable and test output more descriptive.

- Replace repetitive test loops with @pytest.mark.parametrize
- Use descriptive test IDs for better test output
- Group test cases by type with clear comments
- Make parameter names more explicit (path_generator, test_case_description)

AlePouroullis committed:
fix(test): use proper typing.Callable for path generator
